src/main/java/org/killbill/billing/plugin/vertex/VertexTaxCalculator.java

```diff
@@ -305,6 +313,56 @@ private Collection<InvoiceItem> toInvoiceItems(final UUID invoiceId,
         }
     }
 
+    private InvoiceItem createTaxInvoiceItem(final InvoiceItem taxableItem, final UUID invoiceId, @Nullable final InvoiceItem adjustmentItem, final BigDecimal calculatedTax, @Nullable final String description, @Nullable Double taxRate) {
+        final InvoiceItem taxItem = buildTaxItem(taxableItem, invoiceId, adjustmentItem, calculatedTax, description);
+        if (taxItem == null || taxRate == null) {
+            return taxItem;
+        }
```

Comment:
should we return null in above case. Also, shouldn't we also check if taxrate = 0, we may want to skip creating taxitem in that case as well.

Reply:
Creating/not creating of the taxItem base on existence of the effective rate will change existing behaviour and I am not sure if this is correct to do so, I wanted to leave it as is.

Also I don't think we need to check taxRate on 0 and control tax item creation as it is already controlled by tax amount in the existing code

Returning taxItem here is returning basic (not derived taxItem)
```
 final InvoiceItem taxItem = createTaxInvoiceItem(taxableItem, invoiceId, adjustmentItem, calculatedTax, description, transactionLineDetailModel.getEffectiveRate());
                if (taxItem != null) {
                    invoiceItems.add(taxItem);
                }
```

Reply:
And the basic taxItem will be null in case of not positive amount:

```
 if (amount != null && BigDecimal.ZERO.compareTo(amount) != 0) {
            return adjustmentItem != null ? PluginInvoiceItem.createTaxItem(originalItem, invoiceId, adjustmentItem.getStartDate(), adjustmentItem.getEndDate(), amount, (String)MoreObjects.firstNonNull(description, "Tax")) : PluginInvoiceItem.createTaxItem(originalItem, invoiceId, amount, (String)MoreObjects.firstNonNull(description, "Tax"));
        } else {
            return null;
        }
```

Reply:
based on above, tax rate should never be null for `taxItem` as the amount is non-zero. So, we don't even need that additional` || taxRate == null` check

Reply:
taxRate can be theoretically be null (and we don't control if at some point there will be a bug on vertex side), hence I checked on that to not to construct derived tax item without root information why it is being created (taxRate). Otherwise we will return taxDetails = {"taxRate": null} or same taxItem but created as duplicate (with taxRate as null value if there will be an exception during serialization). I don't see the problem to check on null and make life easier here.